target/arc/mmu-v6.c

```diff
@@ -415,30 +347,76 @@ arc_mmuv6_aux_set(const struct arc_aux_reg_detail *aux_reg_detail,
 
 #define ALL1_64BIT (0xffffffffffffffff)
 
+/* TODO: This is for MMU48 only. */
+static char x_for_ttbc(CPUARCState *env, unsigned char i)
+{
+    uint32_t tnsz;
+    const char xs[MMUV6_VERSION_SIZE][2] = {
+        [MMUV6_52_64K] = { 13, 16 },
+        [MMUV6_48_4K]  = { 12, 12 },
+        [MMUV6_48_16K] = {4, 6 },
+        [MMUV6_48_64K] = { 9, 13 },
+    };
+
+    switch (i) {
+    case 0:
+        tnsz = FIELD_EX32(env->mmu.v6.ttbcr, MMU_TTBC, T0SZ);
+        break;
+    case 1:
+        tnsz = FIELD_EX32(env->mmu.v6.ttbcr, MMU_TTBC, T1SZ);
+        break;
+    default:
+        assert(0);
+        break;
+    }
+
+    switch(mmu_v6_version->id) {
+    case MMUV6_32_4K:
+        if(tnsz > 1)
+            return (14 - tnsz);
+        else
+            return (5 - tnsz);
+        break;
+    default:
+        return xs[mmu_v6_version->id][i];
+        break;
+    }
+
+    assert("This should not happen !!!" == 0);
+}
+
+#define MASK_FOR_ROOT_ADDRESS(X) \
+    (((1ull << VADDR_SIZE()) - 1) & (~((1 << X) - 1)))
+
+#define MMU_RTPN_ROOT_ADDRESS(N) \
+    (root_address(env->mmu.v6.rtp##N) & MASK_FOR_ROOT_ADDRESS(x_for_ttbc(env, N)))
+
 static uint64_t
-root_ptr_for_vaddr(uint64_t vaddr, bool *valid)
+root_ptr_for_vaddr(CPUARCState *env, uint64_t vaddr, bool *valid)
```

Comment:
Please add a comment about what this function should do.

Reply:
The same as for the previous request.